procfs_hw/procfs_hw.c

```diff
@@ -0,0 +1,123 @@
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/slab.h>
+#include <linux/string.h>
+#include <linux/timer.h>
+#include <linux/major.h>
+#include <linux/fs.h>
+#include <linux/err.h>
+#include <linux/ioctl.h>
+#include <linux/init.h>
+#include <linux/proc_fs.h>
+#include <asm-generic/uaccess.h>
+#include <linux/types.h>
+#include <linux/bug.h>
+#include <linux/io.h>
+#include <asm/page.h>
+
+
+
+#ifndef __GFP_WAIT
+#define __GFP_WAIT __GFP_RECLAIM
+#endif
+
+#ifndef GFP_IOFS
+#define GFP_IOFS STP_ALLOC_FLAGS
+#endif
+
+#ifdef CONFIG_PROC_FS
+
+#define procfs_msg "Hello World\n"
+#define procfs_name "procfs_hello_world"
+#define procfs_perms 0644
+#define procfs_parent NULL
+
+static struct file_operations procfs_fops;
+static char    *message;
+static int	read_p;
+
+
+static		ssize_t
+pfs_hw_read(struct file *file,
+	    char __user * buf, size_t count, loff_t * offp)
+{
+	ssize_t		len = strlen(message);
+
+	read_p = !read_p;
+	if (read_p) {
+		return 0;
+	}
+	printk("proc called read %ld\n", count);
+	printk("proc called len %ld\n", len);
+	printk("about to copy %s\n", message);
+
+	/*
+	 * should be len instead of 0, but that causes an oops
+	 */
+
+	/* @todo figure out why copy_to_user so upset wich nonzero */
+	copy_to_user(buf, message, 0);
```

Comment:
this should be 
copy_to_user(buf, message, len)
I'm not sure if it is my vm that is bad or a real bug.